src/Storages/System/StorageSystemZooKeeper.cpp

```diff
@@ -459,75 +460,91 @@ void ReadFromSystemZooKeeper::fillData(MutableColumns & res_columns) const
+                    return (task.path_part + '/' + node).substr(0, task.prefix.size()) != task.prefix;
                 });
             }
 
-            std::vector<std::future<Coordination::GetResponse>> futures;
-            futures.reserve(nodes.size());
+            get_tasks.reserve(get_tasks.size() + nodes.size());
             for (const String & node : nodes)
-                futures.push_back(zookeeper->asyncTryGet(path_part + '/' + node));
-
-            for (size_t i = 0, size = nodes.size(); i < size; ++i)
             {
-                context->getProcessListElement()->checkTimeLimit();
-                auto res = futures[i].get();
-                if (res.error == Coordination::Error::ZNONODE)
-                    continue; /// Node was deleted meanwhile.
-
-                // Deduplication
-                String key = path_part + '/' + nodes[i];
-                if (auto [it, inserted] = added.emplace(key); !inserted)
-                    continue;
+                paths_to_get.emplace_back(task.path_part + '/' + node);
+                get_tasks.emplace_back(GetTask{list_task_idx, node});
+            }
+        }
 
-                const Coordination::Stat & stat = res.stat;
-
-                size_t col_num = 0;
-                res_columns[col_num++]->insert(nodes[i]);
-                res_columns[col_num++]->insert(res.data);
-                res_columns[col_num++]->insert(stat.czxid);
-                res_columns[col_num++]->insert(stat.mzxid);
-                res_columns[col_num++]->insert(UInt64(stat.ctime / 1000));
-                res_columns[col_num++]->insert(UInt64(stat.mtime / 1000));
-                res_columns[col_num++]->insert(stat.version);
-                res_columns[col_num++]->insert(stat.cversion);
-                res_columns[col_num++]->insert(stat.aversion);
-                res_columns[col_num++]->insert(stat.ephemeralOwner);
-                res_columns[col_num++]->insert(stat.dataLength);
-                res_columns[col_num++]->insert(stat.numChildren);
-                res_columns[col_num++]->insert(stat.pzxid);
-                res_columns[col_num++]->insert(
-                    task.path); /// This is the original path. In order to process the request, condition in WHERE should be triggered.
-
-                if (task.path_type != ZkPathType::Exact && res.stat.numChildren > 0)
-                {
-                    paths.emplace_back(key, ZkPathType::Recurse);
-                }
+        auto get_responses = zookeeper->tryGet(paths_to_get);
+
+        for (size_t i = 0, size = get_tasks.size(); i < size; ++i)
+        {
+            auto & res = get_responses[i];
+            if (res.error == Coordination::Error::ZNONODE)
+                continue; /// Node was deleted meanwhile.
+
+            auto & get_task = get_tasks[i];
+            auto & list_task = list_tasks[get_task.list_task_idx];
+            context->getProcessListElement()->checkTimeLimit();
```

Comment:
It's possible to run it from a background operation, so `context->getProcessListElement()` will be nullptr

Reply:
https://pastila.nl/?00170ec1/4e423038e8e2d0f12b2806aa3910e38b#+nBHJqZPSOArPKUFv7LIMw==